jormungandr/src/blockchain/storage.rs

```diff
@@ -310,246 +193,70 @@ impl Storage {
-        loop {
-            // Avoid polling on the costly pump machinery while we can expect
-            // the stream to produce values.
-            if self.pressure == 0 {
-                match self.as_mut().poll_pump(cx) {
-                    Poll::Pending => {
-                        return self.as_mut().stream().poll_next(cx);
-                    }
-                    Poll::Ready(()) => {}
-                }
-            } else {
-                match self.as_mut().stream().poll_next(cx) {
-                    Poll::Ready(Some(item)) => {
-                        *self.as_mut().pressure() -= 1;
-                        return Some(item).into();
-                    }
-                    Poll::Ready(None) => return None.into(),
-                    Poll::Pending => {
-                        ready!(self.as_mut().poll_pump(cx));
-                    }
-                }
-            }
-        }
-    }
-}
-
-impl<T, F> BlockIterState<T, F>
-where
-    F: FnMut(Block) -> T,
-{
-    fn new(to_info: BlockInfo<HeaderHash>, distance: u64, transform: F) -> Self {
-        BlockIterState {
-            to_length: to_info.chain_length,
-            cur_length: to_info.chain_length - distance,
-            transform,
-            pending_infos: vec![to_info],
-            pending_item: None,
-        }
-    }
-
-    fn has_next(&self) -> bool {
-        self.cur_length < self.to_length
-    }
-
-    // Iterates the blocks accordingly to this iterator's properties
-    // and sends them to the intercom channel until
-    // the iteration is complete, the channel is full, or an error occurs.
-    // If a storage error is encountered, it is also sent to the channel,
-    // after which iteration terminates.
-    fn fill_sink(
-        mut self: Box<Self>,
-        store: &mut NodeStorageConnection,
-        sink: &mut ReplyStreamSink<T>,
-    ) -> Result<BlockIteration<T, F>, ReplySendError> {
-        if let Some(item) = self.pending_item.take() {
-            let is_err = item.is_err();
-            if !self.try_send_item(item, sink)? {
-                return Ok(BlockIteration::Continue(self));
-            } else if is_err {
-                return Ok(BlockIteration::Break);
-            }
-        }
-        while self.has_next() {
-            match self.get_next_block(store) {
-                Ok(block) => {
-                    let content = (self.transform)(block);
-                    if !self.try_send_item(Ok(content), sink)? {
-                        return Ok(BlockIteration::Continue(self));
+        let mut ancestor = None;
+        let mut closest_found = std::u32::MAX;
+
+        for checkpoint in checkpoints {
+            // Checkpoints sent by a peer may not
+            // be present locally, so we need to ignore certain errors
+            match self
+                .storage
+                .is_ancestor(checkpoint.as_bytes(), descendant.as_bytes())
```

Comment:
Now that it is a blocking operation, what's the worst case performance for it?

I think this whole operation should be lowered into the storage implementation and optimized: can we limit it to one walk through descendant's parents in the volatile store, comparing each level's block hash to all checkpoints (possibly collected in a `HashSet`) at once?

Reply:
The worst performance will be O(n) where n is the number of blocks in the branch of the volatile storage. But this will depend on how often we remove older branches, because if there is no branching at the length of the given block, the performance will be O(1).

Doing what you suggested makes sense in general, but how often will we hit case with the O(n) performance?

Reply:
No idea, but the number of checkpoints that a node sends in a pull request is supposed to grow logarithmically with the length of the current branch in volatile store. Let's optimize this later.